### ISSUE TITLE
pattern DOM-property based on regex works very strange

### ISSUE THREAD

**Issue by abakumov-v** [andyhu92/react-bootstrap4-form-validation#2]
Hello!

I have component `OperationAmountInput`:
```js
import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Label, Input } from 'reactstrap'
import { ValidationForm, TextInput, SelectGroup, Checkbox, Radio, BaseFormControl } from 'react-bootstrap4-form-validation'
import { max20DigitsWithPrecision2Regex } from 'constants/regex'

const operationAmountValueErrorMessage = 'Проверьте, корректно ли указана сумма дохода - правильно: "1200", "500.15", "430,20" (после запятой/точки - максимум 2 символа)'
const operationAmountRequiredErrorMessage = 'Укажите сумму операции'

class OperationAmountInput extends Component {
    constructor(props) {
        super(props);
    }

    handleChangeOperationAmount = (e) => {
        const { onChange } = this.props
        if (onChange)
            onChange(e.target.value)
    }

    render() {
        const { id, selectedValue } = this.props
        const title = `${operationAmountRequiredErrorMessage}. После ввода - ${operationAmountValueErrorMessage.toLowerCase()}`
        return (
            <div>
                <Label for={id} hidden>Укажите сумму операции</Label>
                <TextInput type='text' name='text' id={id} placeholder='Сумма операции'
                    value={selectedValue}
                    onChange={this.handleChangeOperationAmount}
                    required
                    pattern={max20DigitsWithPrecision2Regex}
                    //pattern="^[0-9]{0,20}((\,|\.)\d{0,2})?$"
                    errorMessage={{
                        required: operationAmountRequiredErrorMessage,
                        pattern: operationAmountValueErrorMessage,
                    }}
                />
            </div>
        );
    }
}

OperationAmountInput.propTypes = {
    id: PropTypes.string,
    selectedValue: PropTypes.string,

    onChange: PropTypes.func,
};

export default OperationAmountInput;
```

The `TextInput` has `pattern` property which is taken from the file `regex.js`with next code:
```js
export const max20DigitsWithPrecision2Regex = "^[0-9]{0,20}((\,|\.)\d{0,2})?$"
export const emailRegex = "^([^@]+)@([\\w\\-\\.]+)\\.(\\w{2,4})$"
```

When I'm try input invalid values I'm get next results:
1) input "s" - field is valid, but this value is actually wrong:
![image](https://user-images.githubusercontent.com/6594815/43221072-67e650ec-9054-11e8-89fa-903ddf11877e.png)
2) input "s2" - field is invalid - this it right:
![image](https://user-images.githubusercontent.com/6594815/43221105-878c98d4-9054-11e8-95f9-4c3efd82d4ea.png)
3) input "123" - field is valid - this is right:
![image](https://user-images.githubusercontent.com/6594815/43221119-933e3732-9054-11e8-83e2-97803a8fc45a.png)
4) input "123.0" - field is invalid - but this is actually right:
![image](https://user-images.githubusercontent.com/6594815/43221139-a39622fc-9054-11e8-9b7b-17aba718638b.png)

You can check my regex this: https://regex101.com/r/CYTvuR/1

What I'm doing wrong?

P.S. If I do this:
```js
<TextInput type='text' name='text' id={id} placeholder='Сумма операции'
    value={selectedValue}
    onChange={this.handleChangeOperationAmount}
    required    
    pattern="^[0-9]{0,20}((\,|\.)\d{0,2})?$"
    errorMessage={{
      required: operationAmountRequiredErrorMessage,
      pattern: operationAmountValueErrorMessage,
     }}
/>
```
then I'm getting error in browser console:
```
Pattern attribute value ^[0-9]{0,20}((\,|\.)\d{0,2})?$ is not a valid regular expression: Uncaught SyntaxError: Invalid regular expression: /^[0-9]{0,20}((\,|\.)\d{0,2})?$/: Invalid escape
```

**Comment by andyhu92** [comment]
Hello again! Let's solve this issue first: `Pattern attribute value ^[0-9]{0,20}((\,|\.)\d{0,2})?$ is not a valid regular expression: Uncaught SyntaxError: Invalid regular expression: /^[0-9]{0,20}((\,|\.)\d{0,2})?$/: Invalid escape`.

When you directly putting regex string like `pattern="^[0-9]{0,20}((\,|\.)\d{0,2})?$"`, according to [this](https://stackoverflow.com/questions/41409872/invalid-escape-in-pattern-html-javascript) SO answer, you can use `^[0-9]{0,20}((,|\.)\d{0,2})?$` or simply use `[0-9]{0,20}((,|\.)\d{0,2})?` as your pattern. Notice I removed the `\` before `,` to resolve the issue, see details in the SO answer.

And when you set regex like this `pattern={max20DigitsWithPrecision2Regex}`, the [JSX renderer](https://reactjs.org/docs/introducing-jsx.html#jsx-prevents-injection-attacks) removed all slashes because `\` is an escape character in strings, and if you inspect the input element, the pattern become `pattern="^[0-9]{0,20}((,|.)d{0,2})?$"`, notice all slashes got removed. So you have to escape `\` in your regex and changed to something like `"^[0-9]{0,20}((,|\\.)\\d{0,2})?$"` 

In summary, if you want to reuse the same regex string, escape all slashes, or if you just want to use it once, checked that SO answer make sure you don't have any unnecesarry escaped characters.

**Comment by abakumov-v** [comment]
@andyhu92 thanks for explanation!
